swaybar/tray/dbusmenu.c

```diff
@@ -0,0 +1,1367 @@
+		for (int i = 0; i < menu->items->length; ++i) {
+			struct swaybar_dbusmenu_menu_item *item = menu->items->items[i];
+			struct swaybar_dbusmenu_menu *child_menu = item->submenu;
+			if (child_menu && child_menu->item_id != 0) {
+				swaybar_dbusmenu_menu_destroy(item->submenu);
+			}
+			free(item->label);
+			free(item->icon_name);
+			free(item->icon_data);
+			free(item);
+		}
+	}
+	list_free(menu->items);
+	free(menu);
+}
+
+void swaybar_dbusmenu_destroy(struct swaybar_dbusmenu *menu) {
+	if (!menu) {
+		return;
+	}
+
+	menu->sni->tray->menu = NULL;
+	menu->sni->tray->menu_pointer_focus = NULL;
+
+	close_menus(menu->menu);
+	swaybar_dbusmenu_menu_destroy(menu->menu);
+	free(menu);
+}
+
+static void xdg_popup_done(void *data, struct xdg_popup *xdg_popup) {
+	struct swaybar_dbusmenu_menu *menu = data;
+	swaybar_dbusmenu_destroy(menu->dbusmenu);
+}
+
+static const struct xdg_popup_listener xdg_popup_listener = {
+	.configure = xdg_popup_configure, .popup_done = xdg_popup_done};
+
+static struct swaybar_dbusmenu_menu_item *
+find_item_under_menu(struct swaybar_dbusmenu_menu *menu, int item_id) {
+	if (!menu->items) {
+		return NULL;
+	}
+
+	for (int i = 0; i < menu->items->length; ++i) {
+		struct swaybar_dbusmenu_menu_item *item = menu->items->items[i];
+		if (item->id == item_id) {
+			return item;
+		}
+		if (item->submenu && item->submenu->item_id != 0) {
+			struct swaybar_dbusmenu_menu_item *found_item =
+				find_item_under_menu(item->submenu, item_id);
+			if (found_item) {
+				return found_item;
+			}
+		}
+	}
+
+	return NULL;
+}
+
+static struct swaybar_dbusmenu_menu_item *
+find_item(struct swaybar_dbusmenu *dbusmenu, int item_id) {
+	if (!dbusmenu->menu) {
+		return NULL;
+	}
+
+	return find_item_under_menu(dbusmenu->menu, item_id);
+}
+
+static struct swaybar_dbusmenu_menu *
+find_parent_menu_under_menu(struct swaybar_dbusmenu_menu *menu,
+		struct swaybar_dbusmenu_menu *child_menu) {
+	if (!menu->items) {
+		return NULL;
+	}
+
+	for (int i = 0; i < menu->items->length; ++i) {
+		struct swaybar_dbusmenu_menu_item *item = menu->items->items[i];
+		struct swaybar_dbusmenu_menu *maybe_child_menu = item->submenu;
+		if (maybe_child_menu && maybe_child_menu->item_id != 0) {
+			if (maybe_child_menu == child_menu) {
+				return menu;
+			}
+			maybe_child_menu = find_parent_menu_under_menu(maybe_child_menu, child_menu);
+			if (maybe_child_menu) {
+				return maybe_child_menu;
+			}
+		}
+	}
+
+	return NULL;
+}
+
+static struct swaybar_dbusmenu_menu *
+find_parent_menu(struct swaybar_dbusmenu_menu *menu) {
+	if (menu && menu->item_id == 0) {
+		return NULL;
+	}
+	struct swaybar_dbusmenu_menu *root_menu = menu->dbusmenu->menu;
+	return find_parent_menu_under_menu(root_menu, menu);
+}
+
+static bool is_in_hotspot(struct swaybar_dbusmenu_hotspot *hotspot, int x, int y) {
+	if (!hotspot) {
+		return false;
+	}
+
+	if (hotspot->x <= x && x < hotspot->x + hotspot->width && hotspot->y <= y &&
+			y < hotspot->y + hotspot->height) {
```

Comment:
```suggestion
	if (hotspot->x <= x && x < hotspot->x + hotspot->width && hotspot->y <= y &&
			y < hotspot->y + hotspot->height) {
```
```suggestion
	if (hotspot->x <= x && x - hotspot->x < hotspot->width && hotspot->y <= y &&
			y - hotspot->y < hotspot->height) {
```

prevent integer overflow